xrdp/xrdp_mm.c

```diff
@@ -1192,24 +1094,233 @@ dynamic_monitor_data(intptr_t id, int chan_id, char *data, int bytes)
+                return advance_error(error, mm->resize_state_machine, description);
+            }
+
+            error = xrdp_wm_load_static_pointers(wm);
+            if (error != 0)
+            {
+                LOG(LOG_LEVEL_INFO, "process_dynamic_monitor_description: xrdp_wm_load_static_pointers failed %d", error);
+                return advance_error(error, mm->resize_state_machine, description);
+            }
+            /* resize the main window */
+            error = xrdp_bitmap_resize(wm->screen, description->description.session_width, description->description.session_height);
+            if (error != 0)
+            {
+                LOG(LOG_LEVEL_INFO, "process_dynamic_monitor_description: xrdp_bitmap_resize failed %d", error);
+                return advance_error(error, mm->resize_state_machine, description);
+            }
+            wm->client_info->display_sizes = description->description;
+
+            advance_resize_state_machine(mm->resize_state_machine, description, WMRZ_ENCODER_CREATE);
+            break;
+        case WMRZ_ENCODER_CREATE:
+            if (mm->encoder == NULL)
+            {
+                mm->encoder = xrdp_encoder_create(mm);
+            }
+            advance_resize_state_machine(mm->resize_state_machine, description, WMRZ_SERVER_INVALIDATE);
+            break;
+        case WMRZ_SERVER_INVALIDATE:
+            /* redraw */
+            error = xrdp_bitmap_invalidate(wm->screen, 0);
+            if (error != 0)
+            {
+                LOG(LOG_LEVEL_INFO, "process_dynamic_monitor_description: xrdp_bitmap_invalidate failed %d", error);
+                return advance_error(error, mm->resize_state_machine, description);
+            }
+            if (module != 0)
+            {
+                // Ack all frames to speed up resize.
+                module->mod_frame_ack(module, 0, INT_MAX);
+            }
+            advance_resize_state_machine(mm->resize_state_machine, description, WMRZ_COMPLETE);
+            break;
+        default:
+            break;
+    }
+    return 0;
+}
+
+/******************************************************************************/
+static int
+dynamic_monitor_process_queue(struct xrdp_mm *self)
+{
+    struct dynamic_monitor_description *description;
+
+    LOG_DEVEL(LOG_LEVEL_TRACE, "dynamic_monitor_process_queue:");
+
+    if (self == 0)
+    {
+        return 0;
+    }
+    if (self->wm->login_state != WMLS_CLEANUP && self->wm->login_state != WMLS_INACTIVE)
+    {
+        LOG(LOG_LEVEL_INFO, "dynamic_monitor_process_queue: Not trying to resize while logging in! State was %s", wm_login_state_to_str(self->wm->login_state));
+        return 0;
+    }
+    if (self->resize_queue == 0)
+    {
+        // Potentially the client does not support resizing, so the queue was never created.
+        return 0;
+    }
+    if (self->resize_queue->count <= 0)
+    {
+        return 0;
+    }
+
```

Comment:
The above test isn't needed, as it's done by the `while()` loop below.

Reply:
Fixed.